### PR TITLE
Add OIDC Pluggin support

### PR DESCRIPTION
**Pull request by zaiste**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -19,14 +19,17 @@
   "author": "Saleor Team",
   "license": "BSD-3-Clause",
   "dependencies": {
+    "cookie": "^0.5.0",
     "graphql": "^16.6.0",
-    "graphql-tag": "^2.12.6"
+    "graphql-tag": "^2.12.6",
+    "js-cookie": "^3.0.5"
```
Comment (mmiszy) [inline package.json:25]: This isn't used anywhere?

File: src/SaleorAuthClient.ts [hunk 2]
```diff
@@ -179,6 +184,7 @@ export class SaleorAuthClient {
   signOut = () => {
     this.accessToken = null;
     this.storageHandler?.clearAuthStorage();
+    document.cookie = "token=; expires=Thu, 01 Jan 1970 00:00:00 UTC; path=/;";
```
Comment (mmiszy) [inline src/SaleorAuthClient.ts:187]: You can use `cookie.serialize` instead:
```suggestion
    document.cookie = cookie.serialize("token", "", { expires: new Date(0) });
```

File: src/SaleorAuthClient.ts [hunk 3]
```diff
@@ -150,6 +151,10 @@ export class SaleorAuthClient {
   fetchWithAuth: Fetch = async (input, init) => {
     const refreshToken = this.storageHandler?.getRefreshToken();
 
+    if (!this.accessToken) {
+      this.accessToken = cookie.parse(document.cookie).token;
```
Comment (mmiszy) [inline src/SaleorAuthClient.ts:155]: Let's also delete the cookie here.
```suggestion
      this.accessToken = cookie.parse(document.cookie).token;
      document.cookie = cookie.serialize("token", "", { expires: new Date(0) });
```

File: src/SaleorExternalAuth.ts [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+type RedirectData = {
+  code: string;
+  state: string;
+};
+
+export type ExternalObtainAccessToken = {
+  token: String;
+  refreshToken: String;
+  csrfToken: String;
```
Comment (mmiszy) [inline src/SaleorExternalAuth.ts:19]:
```suggestion
  token: string;
  refreshToken: string;
  csrfToken: string;
```

File: src/SaleorExternalAuth.ts [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+  | { errors: any[] };
+
+export class SaleorExternalAuth {
+  saleorURL: string;
+  provider: ExternalProvider;
+
+  constructor(saleorURL: string, provider: ExternalProvider) {
+    this.saleorURL = saleorURL;
+    this.provider = provider;
+  }
```
Comment (mmiszy) [inline src/SaleorExternalAuth.ts:34]:
```suggestion
  constructor(private saleorURL: string, private provider: ExternalProvider) {
  }
```

File: src/SaleorExternalAuth.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+      console.error(result.errors[0].message);
+      return null;
+    }
+
+    return result.data;
+  }
+
+  async initiate({ redirectURL }: { redirectURL: string }) {
+    const {
+      externalAuthenticationUrl: { authenticationData: data, errors },
```
Comment (mmiszy) [inline src/SaleorExternalAuth.ts:54]: We should probably handle `errors` here?

File: src/SaleorExternalAuth.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    if ("errors" in result) {
+      console.error(result.errors[0].message);
+      return null;
+    }
+
+    return result.data;
+  }
+
+  async initiate({ redirectURL }: { redirectURL: string }) {
```
Comment (mmiszy) [inline src/SaleorExternalAuth.ts:52]: Nit picking: This function and `obtainAccessToken` are inconsistent. They both accept graphql variables but `initiate` destructures it and `obtainAccessToken` doesn't.

File: src/SaleorExternalAuth.ts [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+  constructor(saleorURL: string, provider: ExternalProvider) {
+    this.saleorURL = saleorURL;
+    this.provider = provider;
+  }
+
+  async makePOSTRequest(query: string, variables: unknown) {
+    const response = await fetch(this.saleorURL, {
+      ...POSTasJSON,
+      body: JSON.stringify({ query, variables }),
```
Comment (mmiszy) [inline src/SaleorExternalAuth.ts:39]: Let's use `getRequestData` here:

https://github.com/saleor/auth-sdk/blob/79a195f189e8185aeec22d15ccc5f12a8e2ebe70/src/utils.ts#L49-L59

File: src/mutations.ts [hunk 9]
```diff
@@ -59,3 +60,34 @@ export const PASSWORD_RESET = gql`
...
+        email
+      }
+      errors {
+        field
+        code
+        message
+      }
+    }
+  }
+`;
```
Comment (mmiszy) [inline src/mutations.ts:93]: These two queries are strings but all other queries and fragments are `DocumentNode`. Why? It's inconsistent.

File: src/next/handler.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+import type { NextApiRequest, NextApiResponse } from "next";
+
+import { serialize } from "cookie";
+import { SaleorExternalAuth } from "../SaleorExternalAuth";
+
+export const createSaleorExternalAuthHandler =
```
Comment (mmiszy) [inline src/next/handler.ts:6]: We should update `README.md` an add a simple example on how to use `createSaleorExternalAuthHandler`

File: src/react/useSaleorExternalAuth.ts [hunk 11]
```diff
@@ -0,0 +1,46 @@
+import { useState, useEffect } from "react";
+import { SaleorExternalAuth } from "../SaleorExternalAuth";
+import { ExternalProvider } from "../types";
+
+export type SaleorExternalAuthState = {
+  loading: boolean;
+  authURL?: string;
+  error?: string;
+};
```
Comment (mmiszy) [inline src/react/useSaleorExternalAuth.ts:9]:
```suggestion
type SaleorExternalAuthState =
  | { loading: true; authURL?: undefined; error?: undefined }
  | { loading: false; authURL: string; error?: undefined }
  | { loading: false; authURL?: undefined; error: unknown };
```

**Review by mmiszy — CHANGES_REQUESTED** [review]

File: README.md [hunk 12]
```diff
@@ -136,6 +136,65 @@ const response = await signIn({
...
+      <div>
+        <Link href={authURL}>Login</Link> 
+      </div>
+      )
+  } else {
+    return (
+      <div>Something went wrong</div>
+    )
+  }
+}
```
Comment (mmiszy) [inline README.md:184]:
```suggestion
import {
  useSaleorAuthContext,
  useSaleorExternalAuth,
} from "@saleor/auth-sdk/react";
import { ExternalProvider } from "@saleor/auth-sdk";
import Link from "next/link";
import { gql, useQuery } from "@apollo/client";

export default function Home() {
  const {
    loading: isLoadingCurrentUser,
    error,
    data,
  } = useQuery(
    gql`
      query CurrentUser {
        me {
          id
          email
          firstName
          lastName
        }
      }
    `
  );
  const { authURL, loading: isLoadingExternalAuth } = useSaleorExternalAuth({
    saleorURL: "<your Saleor instance>",
    provider: ExternalProvider.OpenIDConnect,
    redirectURL: "http://localhost:5375/api/auth/callback",
  });

  const { signOut } = useSaleorAuthContext();

  if (isLoadingExternalAuth || isLoadingCurrentUser) {
    return <div>Loading...</div>;
  }
  if (data?.me) {
    return (
      <div>
        {JSON.stringify(data)}
        <button onClick={() => signOut()}>Logout</button>
      </div>
    );
  }
  if (authURL) {
    return (
      <div>
        <Link href={authURL}>Login</Link>
      </div>
    );
  }
  return <div>Something went wrong</div>;
}
```

Comment (mmiszy) [reply README.md:184]: (formatted code)

Second question: is http://localhost:5375 correct?

**Review by mmiszy — CHANGES_REQUESTED** [review]

File: README.md (hunk 12 above)
Comment (zaiste) [reply README.md]: @mmiszy I've clarified the `localhost`, but this will be changed once we have the 1st version out

**Pull request merged**